Feature/actor controller

PhoebeMitchell opened the pull request:
## Summary
<!--- Describe the change below, including rationale and design decisions -->

Implements methods in ActorController (excluding SetActiveSpeaker). Created ActorData scriptable object which holds data about each actor. ActorController loads instances of these objects for each actor and passes it to an Actor class which handles animations and sprites for individual actors. ActorData instances are loaded from an ActorDictionary scriptable object which lists all the actors available.

Modifies DirectorActionDecoder so that the onNewLine event is invoked by the methods tasked with carrying out actions, rather than being invoked by the OnNewActionLine method. Added the PLAY_ANIMATION. This works the same way as the EMOTION actions, however it waits for the animation finish before moving on to the next action. Animations without an OnAnimationComplete event will play indefinitely.

Implemented the WAIT action. This will calls start a coroutine on the scene controller which calls an event after a specified time to call the next line. Calling the next line in the dialogue controller will call the onNextLine event which cancels the current wait coroutine.

Created a template animator controller of blank animations. Actors have animator overrides which they use to play their own animations.

## Before/after screenshots and/or animated gif
<!--- Skip this if not applicable -->
![dialoguedemo3](https://user-images.githubusercontent.com/67964179/131255929-709a3f5e-aaeb-49e0-aeb3-a928663458fc.gif)
The above GIF showcases the PLAYANIMATION and WAIT actions.

## Testing instructions
<!--- What steps can be taken to manually verify the changes? -->
Open "Actor- TestScene" and play. Press space to progress through the demo.

## Additional information
<!--- Check any relevant boxes with "x" -->
- `[ ]` Changes UI
- `[x]` Introduces new feature
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:
  <!--- Include any project card, github issue, etc. associated with this PR -->

TooManyMoths committed:
Actor Controller can set actor and change their emotion

TooManyMoths committed:
Actors are loaded from an actor list scriptable object

TooManyMoths committed:
Created demo scene

OnNewActionLine in DirectionActionDecoder now uses a coroutine so it can add a delay before getting the next line if needed. Also adjusted DirectorActionDecoder to have more consistent private modifiers.

TooManyMoths committed:
Fixed skipping WAIT action not stopping coroutine

JuiceGrape commented on Assets/Resources/InkDialogueScripts/ActorTestInk.ink:
```diff
@@ -0,0 +1,23 @@
+Uh... um... Dan I...
+
+&ACTOR:Dan
+&EMOTION:Lean
```
Forgot a &SPEAK:Dan

PhoebeMitchell replied:
Added

TooManyMoths committed:
Update to RuntimeAnimationController

JuiceGrape commented on Assets/Scripts/Interfaces/IActorController.cs:
```diff
@@ -1,5 +1,7 @@
 public interface IActorController
 {
+    bool Animating { get; set; }
```
Please make sure that this is only gettable, not settable publicly. If the system is showing an animation should only ever be set by the ActorController itself

PhoebeMitchell replied:
Done

JuiceGrape commented on Assets/Scripts/ActorController.cs:
```diff
@@ -17,32 +29,60 @@ void Start()
     public void SetActiveActor(string actor)
     {
-        Debug.LogWarning("SetActiveActor not implemented");
+        _activeActor = Resources.Load<ActorData>($"Actors/{actor}");
```
I'd prefer we have a list of these in editor instead of loading them on runtime like this, but I'd like a second opinion on this from a secondary reviewer.

PhoebeMitchell replied:
I also made a version of this where all the actors were listed in a scriptable object. But was unsure whether it was better to load all the actors into memory like that, or just load them as needed.

alipapapupu replied:
Yeah, I agree. Having hard reference to the objects should be used whenever possible. It makes sure the project stays clean, makes it easier to see all the ActorDatas the game is currently using, keeps the build size smaller and is overall just better option to use on this.

bas-mentjot replied:
A scriptable object seems a bit overkill. What about a prefab?

PhoebeMitchell replied:
Changed back to the scriptable object solution.

JuiceGrape commented on Assets/Scripts/ActorController.cs:
```diff
@@ -17,32 +29,60 @@ void Start()
+        _activeActor = Resources.Load<ActorData>($"Actors/{actor}");
+        if (_activeActor == null)
+        {
+            Debug.LogError($"Actor \"{actor}\" could not be found.");
```
Kudos on the error logging though. Happy to see this

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -36,6 +36,12 @@ public void OnNewActionLine(string line)
+        StartCoroutine(PerformActions(action, parameters));
+    }
+
+    private IEnumerator PerformActions(string action, string parameters)
```
Please remove the coroutine functionality from this, as we will probably set this class to be fully separate from unity once we get a script checker in editor.

PhoebeMitchell replied:
Done

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -53,7 +59,8 @@ public void OnNewActionLine(string line)
             case "SHAKESCREEN": ShakeScreen(parameters); break;
             case "SCENE": SetScene(parameters); break;
-            case "WAIT": Wait(parameters); break;
+            case "WAIT_FOR_ANIMATION": yield return StartCoroutine(WaitForAnimation()); break;
```
Please move this functionality to the ActorController

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -53,7 +59,8 @@ public void OnNewActionLine(string line)
             case "SCENE": SetScene(parameters); break;
-            case "WAIT": Wait(parameters); break;
+            case "WAIT_FOR_ANIMATION": yield return StartCoroutine(WaitForAnimation()); break;
+            case "WAIT": yield return StartCoroutine(Wait(parameters)); break;
```
Please move this functionality to the SceneController

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -36,6 +36,12 @@ public void OnNewActionLine(string line)
         string action = actionAndParam[0];
         string parameters = actionAndParam[1];
 
+        StopAllCoroutines();
```
General remark for this file: onActionDone should only be invoked on actions that are performed instantly. Probably best to move it to the functions that also handle the decoding, so external controllers can just call OnNextLine in the dialogueController themselves

PhoebeMitchell replied:
Moved onActionDone to the methods called in the switch statement, and OnNextLine is called directly when animations are finished playing.

JuiceGrape commented on Assets/Scripts/TextDecoder/DirectorActionDecoder.cs:
```diff
@@ -36,6 +36,12 @@ public void OnNewActionLine(string line)
         string parameters = actionAndParam[1];
 
+        StopAllCoroutines();
+        StartCoroutine(PerformActions(action, parameters));
```
General remark: Please revert the changes the formatter made. I'd hate for this file to be re-done fully by every programmer that touches it by the coding style their formatter is set to.
Side note: Turn off your formatter

PhoebeMitchell replied:
Done

JuiceGrape commented on Assets/Scripts/ActorController.cs:
```diff
@@ -1,14 +1,26 @@
+
+    private void Awake()
     {
+        _spriteRenderer = GetComponent<SpriteRenderer>();
```
Get the active gameobject representing the actor from the scene controller, and then apply actor changes an animations to that instead, so we can have multiple gameobjects in parallel, with only the one being shown being active,

PhoebeMitchell replied:
ActorController now has an ActiveActor property which can be set, and communicates with an actor class which handles playing animations.

JuiceGrape commented on Assets/Scripts/ActorController.cs:
```diff
@@ -17,32 +29,60 @@ void Start()
     public void ShowActor()
     {
-        Debug.LogWarning("ShowActor not implemented");
+        _spriteRenderer.enabled = true;
```
With the changes to the active gameobject for the actor being set in the scene manager, showing the actor might be better suited for the scene controller instead of the actor controller. Please move it there and then leave it to be implemented when the scene controller gets implemented.

PhoebeMitchell replied:
Done. Also moved HideActor to SceneController, updated interfaces, and updated SetActorVisibility method to reflect the change.

JuiceGrape commented on Assets/Resources/InkDialogueScripts/ActorTestInk.ink:
```diff
@@ -0,0 +1,23 @@
+Watch this!
+
+&EMOTION:AirGuitar
+&WAIT_FOR_ANIMATION:
```
Seeing as this has no parameters, should we maybe change the decoder to handle a function without a ':', or should we make sure this has some parameters? (I think the first one is better)

PhoebeMitchell replied:
Actions without parameters no longer require a ':'

JuiceGrape reviewed (changes requested):
Very good start, but gonna need some changes. As you mentioned, there were one or two things that could have been prevented if I had the scrum board in order, so my apologies for that. 
Kudos for the error handling and catching the "wait" and "wait_for_end_of_animation" functionality though

alipapapupu commented on Assets/Scripts/ActorData.cs:
```diff
@@ -0,0 +1,14 @@
+public class ActorData : ScriptableObject
+{
+    [field: SerializeField, Tooltip("The actor's sprite that will appear in the profiles menu.")]
+    public Sprite Profile { get; private set; }
```
Since these are public, they don't necessarily need the SerializeField. Not wrong to have them, just a remark.

PhoebeMitchell replied:
They are properties so need [field: SerializeField] to be serialized. They are properties to prevent their value being changed by scripts on runtime.

alipapapupu commented on Assets/Scripts/ActorController.cs:
```diff
@@ -17,32 +29,60 @@ void Start()
-
     }
 
     public void ShowActor()
```
Why isn't the method commented properly?

JuiceGrape replied:
General remark for this code

PhoebeMitchell replied:
Added summaries to methods with more than one line.

alipapapupu reviewed (approved):
Overall seems good, didn't really find too many improvements that werent mentioned already.

TooManyMoths committed:
Created ActorAnimator animator controller.

Animator overrides should be used to implement individual character animations. Any number of animations can be added to ActorAnimator.

TooManyMoths committed:
Added Dan demo

TooManyMoths committed:
Reverted DirectorActionDecoder

TooManyMoths committed:
Merge branch 'develop' into feature/Actor-Controller

TooManyMoths committed:
Characters are now loaded on scene load in a scriptable object instead of using Resources.Load

Also correct assembly GG-JointJusticw to GG-JointJustice

TooManyMoths committed:
Created actor class, moved actor related classes to their own folder. Moved ShowActor and HideActor methods to SceneController from ActorController as requested by JuiceGrape

TooManyMoths committed:
Added WAIT_FOR_ANIMATION action to DirectorActionDecoder. Modified it to allow parameterless actions

TooManyMoths committed:
Animations without dialogue can now play without being skipped.

TooManyMoths committed:
Moved some ActorController functionality to Actor to allow for multiple actors in one scene.

TooManyMoths committed:
Merge branch 'feature/28-Actor-talks-in-current-pose-while-text-is-progressing' into feature/Actor-Controller

TooManyMoths committed:
Updated animations to use new animation system (the one just merged in using animator overrides)

TooManyMoths committed:
Added PLAY_ANIMATION and implemented WAIT command

TooManyMoths committed:
Created a new demo ink script: ActorTestInk2.ink

TooManyMoths committed:
Merge branch 'develop' into feature/Actor-Controller

TooManyMoths committed:
Fixed some grammar and saved Actor- TestScene
